File: bridge/ipc.py
"""Galley IPC Protocol v0.1 implementation.

Source of truth for the wire format is `docs/ipc-protocol.md`. Keep them in sync;
change the doc first, then the dataclasses here.

Field names are camelCase to match the JSON wire format directly. This makes
serialization a trivial `dataclasses.asdict` call without any name remapping.
"""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field, is_dataclass
from datetime import datetime, timezone
from typing import Any, cast

PROTOCOL_VERSION = "0.1"


def _now_iso() -> str:
    return datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds")


# ---------------- Events (bridge -> workbench) ----------------


@dataclass
class ReadyEvent:
    sessionId: str
    protocolVersion: str
    gaCommit: str
    # ISO 8601 commit date from `git log -1 --format=%cI`. `"unknown"`
    # when ga_path isn't a git checkout (tarball/zip install). Pairs
    # with gaCommit for Settings → Runtime "GA 版本" row.
    gaCommitDate: str
    gaPath: str
    llmName: str
    cwd: str
    pid: int
    availableLLMs: list[dict[str, Any]] = field(default_factory=list)
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "ready"


@dataclass
class TurnStartEvent:
    sessionId: str
    turnIndex: int
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "turn_start"


@dataclass
class ToolCallPendingEvent:
    sessionId: str
    approvalId: str
    turnIndex: int
    toolName: str
    args: dict[str, Any]
    argsPreview: str
    riskLevel: str  # "low" | "medium" | "high"
    reason: str
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "tool_call_pending"


@dataclass
class ToolCallStartEvent:
    sessionId: str
    toolCallId: str
    turnIndex: int
    toolName: str
    args: dict[str, Any]
    argsPreview: str
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "tool_call_start"


@dataclass
class ToolCallEndEvent:
    sessionId: str
    toolCallId: str
    status: str  # "success" | "failed" | "denied" | "cancelled"
    resultPreview: str
    elapsedMs: int
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "tool_call_end"


@dataclass
class ToolCallProgressEvent:
    sessionId: str
    toolCallId: str
    text: str
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "tool_call_progress"


@dataclass
class TurnEndEvent:
    sessionId: str
    turnIndex: int
    summary: str
    toolCalls: list[dict[str, Any]]
    toolResults: list[dict[str, Any]]
    responseContent: str
    exitReason: dict[str, Any] | None = None
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "turn_end"


@dataclass
class TurnProgressEvent:
    """Streaming partial response from the LLM mid-turn (PRD §13.2,
    DESIGN.md §4.3 streaming generation).

    Bridge subscribes to GA's `display_queue` (the same queue
    `agentmain.put_task` returns and fsapp.py drains) and forwards
    each partial chunk over IPC. `delta` is the new substring since
    the last progress event — desktop accumulates these into
    `inFlightContent` and renders mid-turn so users see the LLM's
    output appearing live rather than after the full turn settles.

    `delta` is GA-raw — still contains <thinking>/<summary>/
    <tool_use>/<file_content> tags. Desktop strips them at render
    time (with robust handling of unclosed tags at the partial's
    tail).
    """

    sessionId: str
    delta: str
    source: str  # GA's source field: "workbench" / "system" / etc
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "turn_progress"


@dataclass
class AskUserEvent:
    sessionId: str
    question: str
    candidates: list[str]
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "ask_user"


@dataclass
class RunCompleteEvent:
    sessionId: str
    exitReason: dict[str, Any]
    finalContent: str
    totalTurns: int
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "run_complete"


@dataclass
class ErrorEvent:
    sessionId: str
    message: str
    # Structured triage fields — see docs/ipc-protocol.md §4.10.
    # All have defaults so existing call sites stay valid; bridge classifies
    # known errors before emitting and desktop renders accordingly.
    category: str = "bridge"  # "bridge" | "runtime" | "business"
    severity: str = "error"  # "error" | "warning" | "info"
    retryable: bool = False
    hint: str | None = None  # "check_llm_config" | "network" | "quota_exceeded" | None
    context: str | None = None
    traceback: str | None = None
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "error"


@dataclass
class HistoryLoadedEvent:
    sessionId: str
    messageCount: int
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "history_loaded"


@dataclass
class LLMChangedEvent:
    sessionId: str
    index: int
    name: str
    displayName: str
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "llm_changed"


@dataclass
class ToolsReinjectedEvent:
    """Bridge confirmation that a `reinject_tools` command succeeded.

    Mirrors GA's stapp.py 'Reinject Tools' button behavior — re-reads
    GA's `assets/tool_usable_history.json` and appends those tool
    definition blocks back into `agent.llmclient.backend.history`, so
    an LLM whose tool understanding drifted in a long session gets a
    fresh reminder of what's available. Bridge emits this event on
    success; failures use the standard `error` event.
    """

    sessionId: str
    blocksAdded: int
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "tools_reinjected"


@dataclass
class PetAttachedEvent:
    """Bridge confirmation that the Desktop Pet subprocess has been
    spawned and the per-turn hook is registered on this session's
    agent. Desktop uses this to flip the menu item's "running" state.
    """

    sessionId: str
    port: int
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "pet_attached"


@dataclass
class PetDetachedEvent:
    """Bridge confirmation that the Desktop Pet subprocess has been
    terminated and the per-turn hook removed.
    """

    sessionId: str
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "pet_detached"


@dataclass
class SystemMessageEvent:
    """A standalone, non-agent-loop message emitted into the
    conversation. Used by GA's slash-command paths (`/btw`,
    `/session.x=v`, etc.) where the response bypasses
    agent_runner_loop and goes straight to `display_queue` with
    `source='system'`. The bridge translates that signal into this
    event so desktop can render the content inline.

    `content` is markdown source; desktop renders through the same
    pipeline as agent final answers.

    `kind` discriminates rendering (yellow side-question callout vs
    a more neutral confirmation pill, etc.) — initial values:
      - "side_question" : `/btw` reply
      - "system"        : generic catch-all (default)
    """

    sessionId: str
    content: str
    variant: str = "system"
    timestamp: str = field(default_factory=_now_iso)
    kind: str = "system_message"


Event = (
    ReadyEvent
    | TurnStartEvent
    | ToolCallPendingEvent
    | ToolCallStartEvent
    | ToolCallEndEvent
    | ToolCallProgressEvent
    | TurnEndEvent
    | TurnProgressEvent
    | AskUserEvent
    | RunCompleteEvent
    | ErrorEvent
    | HistoryLoadedEvent
    | LLMChangedEvent
    | ToolsReinjectedEvent
    | PetAttachedEvent
    | PetDetachedEvent
    | SystemMessageEvent
)


# ---------------- Commands (workbench -> bridge) ----------------


@dataclass
class UserMessageCommand:
    text: str
    images: list[str] = field(default_factory=list)
    kind: str = "user_message"


@dataclass
class ApprovalResponseCommand:
    approvalId: str
    decision: str  # "allow_once" | "deny" | "always_allow_project" | "always_allow_global"
    kind: str = "approval_response"


@dataclass
class AskUserResponseCommand:
    text: str
    kind: str = "ask_user_response"


@dataclass
class AbortCommand:
    kind: str = "abort"


@dataclass
class LoadHistoryCommand:
    messages: list[dict[str, Any]]
    kind: str = "load_history"


@dataclass
class SetApprovalRulesCommand:
    alwaysAllowGlobal: list[str] = field(default_factory=list)
    alwaysAllowProject: list[str] = field(default_factory=list)
    kind: str = "set_approval_rules"


@dataclass
class SetYoloModeCommand:
    """Toggle YOLO mode (PRD §11.5).

    When enabled, every dispatched tool call bypasses the approval
    gate — no `tool_call_pending` is emitted; the tool runs as if
    every approval were `allow_once`. The desktop is expected to
    keep the user informed via the persistent TopBar indicator
    (DESIGN.md §4.1).
    """

    enabled: bool
    kind: str = "set_yolo_mode"


@dataclass
class SetLLMCommand:
    llmIndex: int
    kind: str = "set_llm"


@dataclass
class ShutdownCommand:
    kind: str = "shutdown"


@dataclass
class ReinjectToolsCommand:
    """Re-inject GA's tool definitions into the LLM history of this
    session. Useful when an agent's long-running context has caused
    its tool understanding to drift. Reads from GA's
    `assets/tool_usable_history.json` (read-only — see CLAUDE.md
    constitution §"关于读取" — direct read of GA internal asset is
    allowed because read-only, but path is a coupling point that must
    be re-audited at GA baseline upgrades).
    """

    kind: str = "reinject_tools"


@dataclass
class AttachPetCommand:
    """Spawn GA's `desktop_pet_v2.pyw` subprocess and register a per-
    turn hook on this session's agent to push progress updates to
    the pet's local HTTP listener. The pet is global (only one
    instance can run at a time since the pyw binds a fixed port).
    """

    port: int = 41983
    kind: str = "attach_pet"


@dataclass
class DetachPetCommand:
    """Terminate the desktop pet subprocess (if running) and remove
    the per-turn hook from this session's agent.
    """

    kind: str = "detach_pet"


Command = (
    UserMessageCommand
    | ApprovalResponseCommand
    | AskUserResponseCommand
    | AbortCommand
    | LoadHistoryCommand
    | SetApprovalRulesCommand
    | SetYoloModeCommand
    | SetLLMCommand
    | ShutdownCommand
    | ReinjectToolsCommand
    | AttachPetCommand
    | DetachPetCommand
)


# ---------------- Serialization ----------------


EVENT_KINDS: dict[str, type] = {
    "ready": ReadyEvent,
    "turn_start": TurnStartEvent,
    "tool_call_pending": ToolCallPendingEvent,
    "tool_call_start": ToolCallStartEvent,
    "tool_call_end": ToolCallEndEvent,
    "tool_call_progress": ToolCallProgressEvent,
    "turn_end": TurnEndEvent,
    "turn_progress": TurnProgressEvent,
    "ask_user": AskUserEvent,
    "run_complete": RunCompleteEvent,
    "error": ErrorEvent,
    "history_loaded": HistoryLoadedEvent,
    "llm_changed": LLMChangedEvent,
    "tools_reinjected": ToolsReinjectedEvent,
    "pet_attached": PetAttachedEvent,
    "pet_detached": PetDetachedEvent,
    "system_message": SystemMessageEvent,
}

COMMAND_KINDS: dict[str, type] = {
    "user_message": UserMessageCommand,
    "approval_response": ApprovalResponseCommand,
    "ask_user_response": AskUserResponseCommand,
    "abort": AbortCommand,
    "load_history": LoadHistoryCommand,
    "set_approval_rules": SetApprovalRulesCommand,
    "set_yolo_mode": SetYoloModeCommand,
    "set_llm": SetLLMCommand,
    "shutdown": ShutdownCommand,
    "reinject_tools": ReinjectToolsCommand,
    "attach_pet": AttachPetCommand,
    "detach_pet": DetachPetCommand,
}


class IPCProtocolError(Exception):
    """Raised when an IPC message fails to parse or validate."""


def encode(msg: Any) -> str:
    """Serialize a dataclass event/command to a single JSON line (no trailing newline)."""
    if not is_dataclass(msg) or isinstance(msg, type):
        raise IPCProtocolError(f"Not a dataclass instance: {type(msg).__name__}")
    return json.dumps(asdict(msg), ensure_ascii=False, separators=(",", ":"))


def decode_event(line: str) -> Event:
    """Parse a single JSON Lines string as an Event."""
    return cast(Event, _decode(line, EVENT_KINDS, "event"))


def decode_command(line: str) -> Command:
    """Parse a single JSON Lines string as a Command."""
    return cast(Command, _decode(line, COMMAND_KINDS, "command"))


def _decode(line: str, registry: dict[str, type], label: str) -> Any:
    line = line.strip()
    if not line:
        raise IPCProtocolError(f"Empty {label} line")
    try:
        payload = json.loads(line)
    except json.JSONDecodeError as e:
        raise IPCProtocolError(f"Invalid JSON in {label}: {e.msg}") from e
    if not isinstance(payload, dict):
        raise IPCProtocolError(
            f"{label.capitalize()} must be a JSON object, got {type(payload).__name__}"
        )
    kind = payload.get("kind")
    if not isinstance(kind, str):
        raise IPCProtocolError(f"{label.capitalize()} missing 'kind' field")
    cls = registry.get(kind)
    if cls is None:
        raise IPCProtocolError(f"Unknown {label} kind: {kind!r}")
    try:
        return cls(**payload)
    except TypeError as e:
        # missing required field or unexpected field -> dataclass __init__ raises TypeError
        raise IPCProtocolError(f"Invalid {label} {kind!r}: {e}") from e

File: gui/src/types/ipc.ts
/**
 * IPC Protocol v0.1 — desktop-side TypeScript mirror of bridge/ipc.py.
 *
 * Wire format reference: docs/ipc-protocol.md.
 * Every field uses camelCase to match the JSON Lines payload directly.
 *
 * KEEP IN SYNC: when bridge/ipc.py changes, update this file in the same
 * commit. The protocol doc is the source of truth; both Python dataclasses
 * and these TS types implement it.
 */

export const PROTOCOL_VERSION = "0.1";

// ---------------- Shared shapes ----------------

export interface LLMInfo {
  index: number;
  /** Raw "ClassName/model" from GA */
  name: string;
  /** Runtime-mode display label for UI */
  displayName: string;
  isCurrent: boolean;
}

export interface ToolCall {
  toolName: string;
  args: Record<string, unknown>;
  [key: string]: unknown;
}

export interface ToolResult {
  toolUseId?: string;
  content?: unknown;
  [key: string]: unknown;
}

export interface ExitReason {
  result:
    | "CURRENT_TASK_DONE"
    | "EXITED"
    | "MAX_TURNS_EXCEEDED"
    | "ABORTED"
    | string;
  data: unknown;
}

export interface ConversationMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
  toolCalls?: ToolCall[];
  toolResults?: ToolResult[];
}

// ---------------- Events (bridge → desktop) ----------------

export interface ReadyEvent {
  kind: "ready";
  sessionId: string;
  protocolVersion: string;
  /** Full SHA-1 hash from `git rev-parse HEAD` of the user's GA repo.
   * `"unknown"` when ga_path isn't a git checkout. */
  gaCommit: string;
  /** ISO 8601 commit date of the GA HEAD. `"unknown"` when not a git
   * checkout. Pairs with gaCommit for the Settings → Runtime version
   * row + baseline comparison. */
  gaCommitDate: string;
  gaPath: string;
  llmName: string;
  cwd: string;
  pid: number;
  availableLLMs: LLMInfo[];
  timestamp: string;
}

export interface TurnStartEvent {
  kind: "turn_start";
  sessionId: string;
  turnIndex: number;
  timestamp: string;
}

export interface ToolCallPendingEvent {
  kind: "tool_call_pending";
  sessionId: string;
  approvalId: string;
  turnIndex: number;
  toolName: string;
  args: Record<string, unknown>;
  argsPreview: string;
  riskLevel: "low" | "medium" | "high";
  reason: string;
  timestamp: string;
}

export interface ToolCallStartEvent {
  kind: "tool_call_start";
  sessionId: string;
  toolCallId: string;
  turnIndex: number;
  toolName: string;
  args: Record<string, unknown>;
  argsPreview: string;
  timestamp: string;
}

export interface ToolCallEndEvent {
  kind: "tool_call_end";
  sessionId: string;
  toolCallId: string;
  status: "success" | "failed" | "denied" | "cancelled";
  resultPreview: string;
  elapsedMs: number;
  timestamp: string;
}

export interface ToolCallProgressEvent {
  kind: "tool_call_progress";
  sessionId: string;
  toolCallId: string;
  text: string;
  timestamp: string;
}

export interface TurnEndEvent {
  kind: "turn_end";
  sessionId: string;
  turnIndex: number;
  summary: string;
  toolCalls: ToolCall[];
  toolResults: ToolResult[];
  responseContent: string;
  exitReason: ExitReason | null;
  timestamp: string;
}

/**
 * LLM streaming partial. The bridge subscribes to GA's display_queue
 * and forwards each chunk; desktop accumulates `delta` into
 * `inFlightContent` and re-renders the in-flight reply mid-turn.
 *
 * `delta` is GA-raw — still contains <thinking>/<summary>/
 * <tool_use>/<file_content> tags. Desktop strips them at render
 * time, with robust handling of unclosed tags at the partial's tail.
 */
export interface TurnProgressEvent {
  kind: "turn_progress";
  sessionId: string;
  delta: string;
  source: string;
  timestamp: string;
}

export interface AskUserEvent {
  kind: "ask_user";
  sessionId: string;
  question: string;
  candidates: string[];
  timestamp: string;
}

export interface RunCompleteEvent {
  kind: "run_complete";
  sessionId: string;
  exitReason: ExitReason;
  finalContent: string;
  totalTurns: number;
  timestamp: string;
}

/**
 * Structured error event. See docs/ipc-protocol.md §4.10 for the contract.
 *
 * Desktop renders by `category`:
 *   - "runtime"  → conversation inline message bubble
 *   - "bridge"   → top-level toast
 *   - "business" → top-level toast
 *
 * `hint` (when present) maps to a tailored Error Card variant per
 * DESIGN.md §6.2:
 *   - "check_llm_config" — auth / api_key issues
 *   - "network"          — timeout / DNS / connection refused
 *   - "quota_exceeded"   — 429 / rate limits
 */
export interface ErrorEvent {
  kind: "error";
  sessionId: string;
  message: string;
  category: "bridge" | "runtime" | "business";
  severity: "error" | "warning" | "info";
  retryable: boolean;
  hint: "check_llm_config" | "network" | "quota_exceeded" | null;
  context: string | null;
  traceback: string | null;
  timestamp: string;
}

export interface HistoryLoadedEvent {
  kind: "history_loaded";
  sessionId: string;
  messageCount: number;
  timestamp: string;
}

export interface LLMChangedEvent {
  kind: "llm_changed";
  sessionId: string;
  index: number;
  name: string;
  displayName: string;
  timestamp: string;
}

/**
 * Bridge confirmation that a `reinject_tools` command succeeded.
 * `blocksAdded` is how many history entries were appended (one per
 * tool definition block read from GA's `tool_usable_history.json`).
 * Desktop surfaces this as a toast.
 */
export interface ToolsReinjectedEvent {
  kind: "tools_reinjected";
  sessionId: string;
  blocksAdded: number;
  timestamp: string;
}

/**
 * Bridge confirmation that a Desktop Pet subprocess has been spawned
 * for this session and the per-turn hook is registered.
 */
export interface PetAttachedEvent {
  kind: "pet_attached";
  sessionId: string;
  port: number;
  timestamp: string;
}

/**
 * Bridge confirmation that the Desktop Pet subprocess has terminated
 * and the per-turn hook has been removed.
 */
export interface PetDetachedEvent {
  kind: "pet_detached";
  sessionId: string;
  timestamp: string;
}

/**
 * Standalone, non-agent-loop conversation message. Emitted by the
 * bridge when GA's slash-command paths (`/btw`, `/session.x=v`) push
 * a `done` payload to display_queue with `source='system'` — those
 * paths bypass agent_runner_loop, so no turn_end carries the reply.
 *
 * `content` is markdown source — desktop renders through the same
 * pipeline as agent final answers but wraps in a callout chrome
 * matching `variant`.
 *
 * `variant`:
 *   - "side_question": `/btw` reply — yellow callout (AskUserBubble
 *     color family)
 *   - "system": catch-all confirmations — neutral muted register
 */
export interface SystemMessageEvent {
  kind: "system_message";
  sessionId: string;
  content: string;
  variant: "side_question" | "system";
  timestamp: string;
}

export type IPCEvent =
  | ReadyEvent
  | TurnStartEvent
  | ToolCallPendingEvent
  | ToolCallStartEvent
  | ToolCallEndEvent
  | ToolCallProgressEvent
  | TurnEndEvent
  | TurnProgressEvent
  | AskUserEvent
  | RunCompleteEvent
  | ErrorEvent
  | HistoryLoadedEvent
  | LLMChangedEvent
  | ToolsReinjectedEvent
  | PetAttachedEvent
  | PetDetachedEvent
  | SystemMessageEvent;

// ---------------- Commands (desktop → bridge) ----------------

export interface UserMessageCommand {
  kind: "user_message";
  text: string;
  images?: string[];
}

export type ApprovalDecision =
  | "allow_once"
  | "deny"
  | "always_allow_project"
  | "always_allow_global";

export interface ApprovalResponseCommand {
  kind: "approval_response";
  approvalId: string;
  decision: ApprovalDecision;
}

export interface AskUserResponseCommand {
  kind: "ask_user_response";
  text: string;
}

export interface AbortCommand {
  kind: "abort";
}

export interface LoadHistoryCommand {
  kind: "load_history";
  messages: ConversationMessage[];
}

export interface SetApprovalRulesCommand {
  kind: "set_approval_rules";
  alwaysAllowGlobal: string[];
  alwaysAllowProject: string[];
}

/**
 * Toggle YOLO mode on the bridge (PRD §11.5).
 *
 * When enabled, every dispatched tool call bypasses the approval gate
 * — bridge does not emit `tool_call_pending`. The desktop is expected
 * to keep the user informed via the persistent TopBar indicator
 * (DESIGN.md §4.1). YOLO is upper-priority over `always_allow_*`
 * lists; toggling it back off does not clear those lists.
 */
export interface SetYoloModeCommand {
  kind: "set_yolo_mode";
  enabled: boolean;
}

export interface SetLLMCommand {
  kind: "set_llm";
  llmIndex: number;
}

export interface ShutdownCommand {
  kind: "shutdown";
}

/**
 * Re-inject GA's tool definitions into this session's LLM history.
 * Mirrors GA's stapp.py "Reinject Tools" button — useful when a
 * long-running session has drifted and the agent is confused about
 * what tools are available.
 */
export interface ReinjectToolsCommand {
  kind: "reinject_tools";
}

/**
 * Spawn GA's `desktop_pet_v2.pyw` subprocess and register a per-turn
 * progress hook on this session's agent. `port` is the local HTTP
 * port the pet listens on (default 41983, matching GA's stapp.py).
 * Only one pet can run at a time globally since it binds a fixed port.
 */
export interface AttachPetCommand {
  kind: "attach_pet";
  port?: number;
}

/**
 * Terminate the Desktop Pet subprocess (if running) and remove the
 * per-turn hook. No-op if no pet is currently attached.
 */
export interface DetachPetCommand {
  kind: "detach_pet";
}

export type IPCCommand =
  | UserMessageCommand
  | ApprovalResponseCommand
  | AskUserResponseCommand
  | AbortCommand
  | LoadHistoryCommand
  | SetApprovalRulesCommand
  | SetYoloModeCommand
  | SetLLMCommand
  | ShutdownCommand
  | ReinjectToolsCommand
  | AttachPetCommand
  | DetachPetCommand;
